src/modules/rover_differential/RoverDifferentialControl/RoverDifferentialControl.cpp

```diff
@@ -0,0 +1,178 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "RoverDifferentialControl.hpp"
+
+#include <mathlib/math/Limits.hpp>
+
+using namespace matrix;
+using namespace time_literals;
+
+RoverDifferentialControl::RoverDifferentialControl(ModuleParams *parent) : ModuleParams(parent)
+{
+	updateParams();
+	_rover_differential_status_pub.advertise();
+	pid_init(&_pid_yaw_rate, PID_MODE_DERIVATIV_NONE, 0.001f);
+	pid_init(&_pid_throttle, PID_MODE_DERIVATIV_NONE, 0.001f);
+	pid_init(&_pid_yaw, PID_MODE_DERIVATIV_NONE, 0.001f);
+}
+
+void RoverDifferentialControl::updateParams()
+{
+	ModuleParams::updateParams();
+	_max_yaw_rate = _param_rd_max_yaw_rate.get() * M_DEG_TO_RAD_F;
+	pid_set_parameters(&_pid_yaw_rate,
+			   _param_rd_yaw_rate_p.get(), // Proportional gain
+			   _param_rd_yaw_rate_i.get(), // Integral gain
+			   0.f, // Derivative gain
+			   1.f, // Integral limit
+			   1.f); // Output limit
+	pid_set_parameters(&_pid_throttle,
+			   _param_rd_p_gain_speed.get(), // Proportional gain
+			   _param_rd_i_gain_speed.get(), // Integral gain
+			   0.f, // Derivative gain
+			   1.f, // Integral limit
+			   1.f); // Output limit
+	pid_set_parameters(&_pid_yaw,
+			   _param_rd_p_gain_yaw.get(),  // Proportional gain
+			   _param_rd_i_gain_yaw.get(),  // Integral gain
+			   0.f,  // Derivative gain
+			   _max_yaw_rate,  // Integral limit
+			   _max_yaw_rate);  // Output limit
+}
+
+void RoverDifferentialControl::computeMotorCommands(const float vehicle_yaw, const float vehicle_yaw_rate,
+		const float vehicle_forward_speed)
+{
+	// Timestamps
+	hrt_abstime timestamp_prev = _timestamp;
+	_timestamp = hrt_absolute_time();
+	const float dt = math::constrain(_timestamp - timestamp_prev, 1_ms, 5000_ms) * 1e-6f;
+
+	// Update differential setpoint
+	_rover_differential_setpoint_sub.update(&_rover_differential_setpoint);
+
+	// Closed loop yaw control (Overrides yaw rate setpoint)
+	if (_rover_differential_setpoint.closed_loop_yaw && PX4_ISFINITE(_rover_differential_setpoint.yaw_setpoint)) {
+		if (fabsf(_rover_differential_setpoint.yaw_setpoint - vehicle_yaw) < FLT_EPSILON) {
```

Comment:
In which case would you have an error less than FLT_EPSILON? Resp. why do you have to handle it?

Reply:
This is the idea: vehicle_yaw is a sensor measurement with noise, this check is to stop the rover from "jittering" when at a standstill.

Reply:
Makes sense, yes.
I added this check as a fix because I had the jittering problem. But maybe the cause was float precission and sensor noise wasn't actually the problem...

Reply:
Maybe we should hard-code the limit to 5° or so, but we can also leave it like that for now.